File: web/frontend/src/lib/api/knowledge.ts
import { getJSON, sendJSON } from './client'

export type TreeNode = {
  name: string
  type: 'dir' | 'file'
  path: string
  children?: TreeNode[]
}

const q = encodeURIComponent

export function getTree(contextId = 'global'): Promise<TreeNode> {
  return getJSON(`/api/knowledge/tree?context_id=${q(contextId)}`)
}

export async function readFile(path: string, contextId = 'global'): Promise<string> {
  const j = await getJSON<{ content: string }>(
    `/api/knowledge/file?path=${q(path)}&context_id=${q(contextId)}`,
  )
  return j.content
}

export function writeFile(path: string, content: string, contextId = 'global'): Promise<void> {
  return sendJSON('/api/knowledge/file', 'PUT', { path, content, context_id: contextId })
}

export function injectNote(
  title: string,
  content: string,
  folder = 'knowledge',
  contextId = 'global',
): Promise<{ path: string }> {
  return sendJSON('/api/knowledge/inject', 'POST', { title, content, folder, context_id: contextId })
}

File: superme_agent/daemon/schemas/knowledge.py
"""Schemas for the knowledge routes (knowledge.py)."""

from __future__ import annotations

from pydantic import BaseModel


class KnowledgeNode(BaseModel):
    """A node in the knowledge tree. `dir` nodes carry `children`; `file` nodes omit it
    (serialized with response_model_exclude_none so files stay `{name, type, path}`)."""
    name: str
    type: str  # "dir" | "file"
    path: str
    children: list[KnowledgeNode] | None = None


class KnowledgeFileResponse(BaseModel):
    path: str
    content: str


class KnowledgeOkResponse(BaseModel):
    """The write/inject result: ok + the (relative) path written."""
    ok: bool
    path: str


class WriteBody(BaseModel):
    path: str
    content: str
    context_id: str = "global"


class InjectBody(BaseModel):
    title: str
    content: str
    folder: str = "knowledge"
    context_id: str = "global"
